Workspaces support (#484)
foundry-rs/starknet-foundry: docs/src/testing/running-tests.md

[SKIP] package_name::yet_another_test
Tests: 3 passed, 1 failed, 2 skipped
```

> ## Running tests in Scarb Workspace

Reviewer: I think we should link scarb docs about workspaces somewhere in this section

For a project structure like this

```shell
$ tree . -L 3
.
├── Scarb.toml
├── crates
│   ├── addition
│   │   ├── Scarb.toml
│   │   ├── src
│   │   └── tests
│   └── fibonacci
│       ├── Scarb.toml
│       └── src
└── src
    └── lib.cairo
$ snforge
Collected 1 test(s) and 1 test file(s)
Running 1 test(s) from hello_workspaces package
[PASS] hello_workspaces::test_simple
Tests: 1 passed, 0 failed, 0 skipped
```

To specify a package to test, pass a `--package package_name` (or `-p package_name` for short) flag, to select the specific package. You can also run `snforge` from the package directory for the same effect.

```shell
$ snforge --package addition
Collected 4 test(s) and 3 test file(s)
Running 1 test(s) from addition package
[PASS] addition::tests::it_works
Running 2 test(s) from tests/nested/test_nested.cairo
[PASS] test_nested::test_two
[PASS] test_nested::test_two_and_two
Running 1 test(s) from tests/test_simple.cairo
[PASS] test_simple::simple_case
Tests: 4 passed, 0 failed, 0 skipped

> ```

Reviewer: This and the output below could be changed a little bit so there are less tests. It's quite hard to read with so many outputs.

You can also pass `--workspace` flag to explicitly run tests for all packages in the workspace.

```shell
$ snforge --workspace
Collected 4 test(s) and 3 test file(s)
Running 1 test(s) from addition package
[PASS] addition::tests::it_works
Running 2 test(s) from tests/nested/test_nested.cairo
[PASS] test_nested::test_two
[PASS] test_nested::test_two_and_two
Running 1 test(s) from tests/test_simple.cairo
[PASS] test_simple::simple_case
Tests: 4 passed, 0 failed, 0 skipped
Running 1 test(s) from fibonacci package
[PASS] fibonacci::tests::it_works
Tests: 1 passed, 0 failed, 0 skipped
Collected 3 test(s) and 2 test file(s)
Running 1 test(s) from hello_workspaces package
[PASS] hello_workspaces::test_simple
Tests: 1 passed, 0 failed, 0 skipped

> ```

Reviewer: I think a separate subsection showcasing how scarb behaves in virtual workspaces could be useful as well.

Reviewer: Yeah, sorry. I meant `snforge`.

[SKIP] package_name::yet_another_test
Tests: 3 passed, 1 failed, 2 skipped
```

## Scarb workspaces support

Forge supports Scarb Workspaces.
To make sure you know how workspaces work,
check Scarb documentation [here](https://docs.swmansion.com/scarb/docs/reference/workspaces.html).

### Workspaces with root package

When running `snforge` in a Scarb workspace with a root package, it will only run tests inside the root package.  

> I.e., for a project structure like this

Reviewer: ```suggestion
For a project structure like this
```

```shell
$ tree . -L 3
.
├── Scarb.toml
├── crates
│   ├── addition
│   │   ├── Scarb.toml
│   │   ├── src
│   │   └── tests
│   └── fibonacci
│       ├── Scarb.toml
│       └── src
└── tests
    └── test.cairo
└── src
    └── lib.cairo
```

only the tests in `./src` and `./tests` folders will be executed.

```shell

$ snforge
Collected 1 test(s) and 1 test file(s) from hello_workspaces package
Running 1 inline test(s)
[PASS] hello_workspaces::test_simple
Tests: 1 passed, 0 failed, 0 skipped
```

To select the specific package to test, pass a `--package package_name` (or `-p package_name` for short) flag.

> You can also run `snforge` from the package directory to achieve the same effect.

Reviewer: I know what you mean but I'm not sure if this is explicit enough.

Running 1 inline test(s)
[PASS] addition::tests::it_works
Running 1 test(s) from tests/test_simple.cairo
[PASS] test_simple::simple_case
Tests: 2 passed, 0 failed, 0 skipped
```

You can also pass `--workspace` flag to run tests for all packages in the workspace.

```shell
$ snforge --workspace
Collected 2 test(s) and 2 test file(s) from addition package
Running 1 inline test(s)
[PASS] addition::tests::it_works
Running 1 test(s) from tests/test_simple.cairo
[PASS] test_simple::simple_case
Tests: 2 passed, 0 failed, 0 skipped

Collected 1 test(s) and 1 test file(s) from fibonacci package
Running 1 inline test(s)
[PASS] fibonacci::tests::it_works
Tests: 1 passed, 0 failed, 0 skipped

Collected 3 test(s) and 2 test file(s) from hello_workspaces package
Running 1 inline test(s)
[PASS] hello_workspaces::test_simple
Tests: 1 passed, 0 failed, 0 skipped
```

`--package` and `--workspace` flags are mutually exclusive, adding both of them to a `snforge` command will result in an error.

### Virtual workspaces

Running `snforge` command in a virtual workspace (a workspace without a root package)
outside any package will by default run tests for all the packages. 

> Using the `--workspace` flag has the same effect.

Reviewer: ```suggestion
It is equivalent to running `snforge` with the `--workspace` flag.
```
